ndcube/meta.py

```diff
@@ -0,0 +1,489 @@
+
+    Parameters
+    ----------
+    meta: dict-like
+        The names and values of metadata.
+
+    key_comments: dict-like, optional
+        Comments associated with any of the above pieces of metadata.
+
+    axes: dict-like, optional
+        The axis/axes associated with the metadata denoted by the keys.
+        Metadata not included are considered not to be associated with any axis.
+        Each axis value must be an iterable of `int`. An `int` itself is also
+        acceptable if the metadata is associated with a single axis.
+        The value of axis-assigned metadata in meta must be same length as
+        number of associated axes (axis-aligned), or same shape as the associated
+        data array's axes (grid-aligned).
+
+    Notes
+    -----
+    **Axis-aware Metadata**
+    There are two valid types of axis-aware metadata: axis-aligned and grid-aligned.
+    Axis-aligned metadata gives one value per associated axis, while grid-aligned
+    metadata gives a value for each data array element in the associated axes.
+    Consequently, axis-aligned metadata has the same length as the number of
+    associated axes, while grid-aligned metadata has the same shape as the associated
+    axes. To avoid confusion, axis-aligned metadata that is only associated with one
+    axis must be scalar or a string. Length-1 objects (excluding strings) are assumed
+    to be grid-aligned and associated with a length-1 axis.
+
+    **Slicing and Rebinning Axis-aware Metadata**
+    Axis-aligned metadata is only considered valid if the associated axes are present.
+    Therefore, axis-aligned metadata is only changed if an associated axis is dropped
+    by an operation, e.g. slicing. In such a case, the value associated with the
+    dropped axes is also dropped and hence lost.  If the axis of a 1-axis-aligned
+    metadata value (scalar) is slicing away, the metadata key is entirely removed
+    from the NDMeta object.
+
+    Grid-aligned metadata is mirrors the data array, it is sliced following
+    the same rules with one exception. If an axis is dropped by slicing, the metadata
+    name is kept, but its value is set to the value at the row/column where the
+    axis/axes was sliced away, and the metadata axis-awareness is removed. This is
+    similar to how coordinate values are transferred to ``global_coords`` when their
+    associated axes are sliced away.
+
+    Note that because rebinning does not drop axes, axis-aligned metadata is unaltered
+    by rebinning. By contrast, grid-aligned metadata must necessarily by affected by
+    rebinning. However, how it is affected depends on the nature of the metadata and
+    there is no generalized solution. Therefore, this class does not alter the shape
+    or values of grid-aligned metadata during rebinning, but simply removes its
+    axis-awareness.  If specific pieces of metadata have a known way to behave during
+    rebinning, this can be handled by subclasses or mixins.
+    """
+
```

Comment:
The ABC wont enforce this (I don't think there's a way) but we should add them here for completeness
```suggestion

    __ndcube_can_slice__: bool
    __ndcube_can_rebin__: bool
    
```